feat: update /interaction/select-menu (#54)
DisnakeDev/guide: guide/docs/interactions/select-menus.mdx

bot = commands.InteractionBot()

@bot.listen()
async def on_ready():
    print(f"Logged in as {bot.user} (ID: {bot.user.id})\n------")

@bot.slash_command()
async def animals(inter: disnake.ApplicationCommandInteraction):
    """Sends a message with our dropdown containing the animals"""

    # create the view with our dropdown object
    view = DropDownView()

    # respond to the interaction with a message and our view
    await inter.response.send_message("What is your favorite type of animal?", view=view)

if __name__ == "__main__":
    bot.run(os.getenv("BOT_TOKEN"))
```

</TabItem>
<TabItem value="decorator" label="decorator.py">

```python
# instead of subclassing `disnake.ui.StringSelect`, this example shows how to use the
# `@disnake.ui.string_select` decorator directly inside the view to create the dropdown
# component.
class AnimalView(disnake.ui.View):
    def __init__(self):
        super().__init__()

        # rather than creating a large decorator, we can create the options for this component here
        # this also works if you wish to pass a list of options to this view to create options dynamically

>         self.animal_callback.options = [

Author: Right.  If the options are static, I normally wouldn't do it this way, but I also thought that I'd show that it does work and would allow the use of dynamic options, even though they weren't being used with this specific example.   I have no issues with committing an altered version with the options inside the decorator

>         self.animal_callback.options = [

Reviewer: That's completely fair. I think I'd still prefer to highlight the decorator-only solution first and foremost, and instead mention that you can also assign the options separately in a comment or info box; it's certainly useful to know regardless, but makes things more complex for beginners.

not available to the other SelectMenu types.{' '}

`channel_types` will allow you to specify the types of channels made available as options, omit to add all available channels:

-   `channel_types=disnake.ChannelType.text` to display only guild text channels as options
-   `channel_types=disanke.ChannelType.voice` to display only guild voice channels as options.

See <DocsLink reference="disnake.ChannelType">disnake.ChannelType</DocsLink> to see more channel types.
:::

### Handling View Timeouts

When a view times out the bot will no longer listen for these events and your users will receive an error `This interaction failed`.

To prevent this you have a couple of options.

1.  Disable the components within the view so that they are no longer interactive.
2.  Remove the view altogether leaving only the original message.

For this example we will disable the component `on_timeout`. However, if you wish to remove the view completely you can pass `view=None` instead of `view=self` as the example below shows.

We'll continue with the original example from above, however I will only be altering the parts that matter.

```python title="viewtimeout.py"
...

> ...

Reviewer: ```suggestion
```

class DropDownView(disnake.ui.View):

    message: disnake.Message  # adding to typehint the future `message` attribute

    def __init__(self):
        # this time we will set the timeout to 30.0s
        super().__init__(timeout=30.0)
        # now let's add the `StringSelect` object we created above to this view
        self.add_item(AnimalDropdown())

    # to handle this timeout, we'll need to override the default `on_timeout` method within the `View``
    async def on_timeout(self):
        # now we will edit the original command response so that it displays the disabled view
        # since `self.children` returns a list of the components attached to this `View` and we want to
        # disable the single component, we can access it with a simple [0] index
        self.children[0].disabled = True
        # now we edit the message with this updated `View`

>         await self.message.edit(view=self)

Author: Easy enough.

...
...
# the only changes we need to make to handle the updated view is to fetch the original response after
# it has been sent.  This is required because `interaction.send` methods do not return a message object
@bot.slash_command()
async def animals(inter: disnake.ApplicationCommandInteraction):
    """Sends a message with our dropdown containing the animals"""

    # create the view with our dropdown object
    view = DropDownView()

    # respond to the interaction with a message and our view
    await inter.response.send_message("What is your favorite type of animal?", view=view)

    # this will add a new `message` attribute to the `DropDownView` that will be edited
    # once the view has timed out
    view.message = await inter.original_response()

> ...

Reviewer: ```suggestion
```

...
```

### Views vs. low-level components

> As an alternative to using `Views`, it is possible to use SelectMenus as low-level components.
> These components do not need to be sent as part of a View and can be sent as is. However, you should take note
> that any each component being sent in this manner must have a `custom_id` explicitly set. It should also be noted
> that components will share listeners. This means `custom_id` should be unique for each component.
> 
> The main advantage of this is that listeners, by nature, are persistent and will remain fully functional over bot reloads. Listeners are are stored in the bot
> strictly once, and are shared by all components. Because of this, the memory footprint will generally be smaller
> than that of an equivalent view.

Reviewer: ```suggestion
As an alternative to using `View`s, it is possible to use Select Menus as low-level components.
These components do not need to be sent as part of a View and can be sent as is.

Note that any component being sent in this manner must have a `custom_id` explicitly set. Component interactions are sent to all listeners,
which means the `custom_id` should be unique for each component to be able to identify the component in your code.

The main advantage of this is that listeners, by nature, are persistent and will remain fully functional over bot reloads. Listeners are stored in the bot
strictly once, and are shared by all components. Because of this, the memory footprint will generally be smaller
than that of an equivalent view.
```

again, minor formatting changes and small details

The example below will be similar to the above examples, however will be executed as a low level component instead.

```python title="low_level_dropdown.py"
@bot.slash_command()
async def animals(inter: disnake.ApplicationCommandInteraction):
    """Sends a message with our dropdown containing the animals"""

    await inter.response.send_message(
        "What is your favorite type of animal?",
        components=[
            disnake.ui.StringSelect(

>                 options=["Dog", "Cat", "Snake", "Gerbil"], custom_id="fav_animal"

Reviewer: ```suggestion
                custom_id="fav_animal",
                options=["Dog", "Cat", "Snake", "Gerbil"],
```

            )
        ],
    )

> # now we create the listener that will handle the users's selection.  Similar to the callback we used above

Reviewer: ```suggestion
# Now we create the listener that will handle the users's selection, similar to the callback we used above
```

For this example we will disable the component `on_timeout`. However, if you wish to remove the view completely you can pass `view=None` (instead of `view=self` like the example below).

We'll continue with the `subclassing.py` example from above, only altering the relevant parts:

```python title="viewtimeout.py"
...

class DropDownView(disnake.ui.View):

    message: disnake.Message  # adding to typehint the future `message` attribute

    def __init__(self):
        # this time we will set the timeout to 30.0s
        super().__init__(timeout=30.0)
        # now let's add the `StringSelect` object we created above to this view
        self.add_item(AnimalDropdown())

    # to handle this timeout, we'll need to override the default `on_timeout` method within the `View``
    async def on_timeout(self):
        # Now we will edit the original command response so that it displays the disabled view.
        # Since `self.children` returns a list of the components attached to this `View` and we want
        # to prevent the components from remaining interactive after the timeout, we can easily loop
        # over its components and disable them.
        for child in self.children:

>             child.disabled = True

Reviewer: ```suggestion
            if isinstance(child, (disnake.ui.Button, disnake.ui.BaseSelect)):
                child.disabled = True
```
For completeness' sake
